playground/tei_bot_v2_util.py

```diff
@@ -0,0 +1,269 @@
+     [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+     [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+     [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+     [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+     [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+     [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]]
+)
+
+
+@dataclass
+class RowConnected:
+    connected_size: int
+    position: List[int]
+
+
+def get_white(mat) -> np.array:
+    mat = np.where(mat != 2, 0, mat)
+    mat = np.where(mat == 2, 1, mat)
+    return mat
+
+
+def get_black(mat) -> np.array:
+    return np.where(mat != 1, 0, mat)
+
+
+def find_connected_under_k(vec, k):
+    size = len(vec)
+    cum_sum_vec = np.zeros(size, dtype="int32")
+    for i in range(k, 0, -1):
+        conv = np.concatenate((np.convolve(vec, np.ones(i, dtype="int32"), "valid") / i, np.zeros(i - 1)))
+        conv = np.where(conv != 1, 0, conv).astype("int32")
+        cum_sum_vec = cum_sum_vec + conv
+    diff = np.diff(cum_sum_vec)
+    diff = np.concatenate((cum_sum_vec[0:1], diff))
+    diff = np.where(diff < 0, 0, diff)
+    return diff
+
+
+def find_connected_1d(vec):
+    first_element_of_connected_n = []
+    if np.sum(vec) == 0:
+        return first_element_of_connected_n
+    vec_copy = vec.copy()
+    zz4 = find_connected_under_k(vec_copy, 6)
+
+    for k in [6, 5, 4, 3, 2, 1]:
+        n_connected_position = np.where(zz4 == k)[0]
+        if len(n_connected_position) >= 0:
+            connected = RowConnected(connected_size=k, position=list(n_connected_position))
+            first_element_of_connected_n.append(connected)
+    return first_element_of_connected_n
+
+
+def find_connected_edge(dic: List[RowConnected]) -> List[RowConnected]:
+    connected_edge = []
+    for x in dic:
+        edges = sum([[y - 1, y + x.connected_size] for y in x.position], [])
+        edges_filtered_distinct = list(set([v1 for v1 in edges if 0 <= v1 <= 18]))
+        edges_filtered_distinct = RowConnected(x.connected_size, edges_filtered_distinct)
+        connected_edge.append(edges_filtered_distinct)
+    return connected_edge
+
+
+def diagonal_order(matrix, row, col):
+    value = []
+    coord = []
+    for line in range(1, (row + col)):
+        start_col = max(0, line - row)
+        count = min(line, (col - start_col), row)
+        jj = []
+        kk = []
+        for j in range(0, count):
+            jj.append(matrix[min(row, line) - j - 1][start_col + j])
+            kk.append([min(row, line) - j - 1, start_col + j])
+        value.append(np.array(jj))
+        coord.append(kk)
+    value_coord = list(zip(value, coord))
+    return value_coord
+
+
+def find_connected_19x19(mat):
+    mat_copy = mat.copy()
+    mat_copy_flip = np.fliplr(mat.copy())
+    position = []
+    diagonal_0 = diagonal_order(mat_copy, 19, 19)
+    diagonal_1 = diagonal_order(mat_copy_flip, 19, 19)
+
```

Comment:
please use more intuitive names rather than `diagonal_0`,`diagonal_1`.

Reply:
in addition, size 19 may be changed according to board config. It would be better to receive as a parameter rather than a fixed number